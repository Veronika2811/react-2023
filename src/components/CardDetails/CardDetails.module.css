.details {
  position: fixed;
  right: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 20px;
  height: 100%;
  max-width: 300px;
  padding: 30px;
  box-shadow: 0px 4px 20px rgba(250, 250, 250, 0.05);
}
