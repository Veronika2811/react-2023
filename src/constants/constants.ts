export const BASE_URL = 'https://rickandmortyapi.com/api/character';

export const DEFAULT_VALUE_PER_PAGE = 20;
export const ADDITIONAL_VALUE_PER_PAGE = 10;

export const LOCAL_STORAGE_KEY = 'Veronika2811-react-2023__searchRequest';

export const DETAILS_URL_PARAMETER_KEY = 'details';
export const PAGE_URL_PARAMETER_KEY = 'page';

export const DEFAULT_PAGE = '1';
