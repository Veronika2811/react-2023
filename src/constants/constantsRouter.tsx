import { createBrowserRouter } from 'react-router-dom';

import App from '../App';
import ErrorPage from '../pages/ErrorPage';
import CharacterDetailsWrapper from '../components/CharacterDetailsWrapper';

const CONSTANTS_ROUTER = createBrowserRouter([
  {
    path: '/',
    element: <App />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: '/',
        element: <CharacterDetailsWrapper />,
      },
    ],
  },
]);

export default CONSTANTS_ROUTER;
