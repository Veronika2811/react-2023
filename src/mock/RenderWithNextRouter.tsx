import React, { ReactNode } from 'react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import { createMockRouter } from './createMockRouter';

interface IRenderWithNextRouter {
  queryParams?: { [key: string]: string };
  children: ReactNode;
}

const RenderWithNextRouter = ({
  queryParams = {},
  children,
}: IRenderWithNextRouter) => {
  const { Provider } = RouterContext;

  console.log(children);

  return (
    <Provider value={createMockRouter({ query: queryParams })}>
      {children}
    </Provider>
  );
};

export default RenderWithNextRouter;
