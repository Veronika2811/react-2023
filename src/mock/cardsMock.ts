export const MortyMock = {
  created: '2017-11-04T18:50:21.651Z',
  episode: ['https://rickandmortyapi.com/api/episode/1'],
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  name: 'Morty Smith',
  origin: {
    name: 'unknown',
    url: '',
  },
  species: 'Human',
  status: 'Alive',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/2',
};

export const AlbertMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 11,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Albert Einstein',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
