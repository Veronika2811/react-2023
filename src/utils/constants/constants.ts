export const BASE_URL = 'https://rickandmortyapi.com/api/character';

export const DEFAULT_PAGE = '1';

export const DEFAULT_VALUE_PER_PAGE = '20';
export const ADDITIONAL_VALUE_PER_PAGE = '10';

export const DEFAULT_QUERY_PARAMS = {
  currentPage: DEFAULT_PAGE,
  perPage: DEFAULT_VALUE_PER_PAGE,
};
