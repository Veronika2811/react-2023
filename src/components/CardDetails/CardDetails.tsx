import { useSearchParams } from 'react-router-dom';

import Preloader from '../UI/preloader/Preloader';
import Button from '../UI/button/Button';
import getStatusCharacterColor from '../../utils/getStatusCharacterColor';
import { useGetCharacterItemQuery } from '../../redux/api/apiSlice';
import { useAppDispatch, useAppSelector } from '../../redux/hooks';
import { charactersChangeViewMode } from '../../redux/store/charactersSlice';
import { RootState } from '../../redux/store/store';
import { DETAILS_URL_PARAMETER_KEY } from '../../constants/constants';

import classes from './CardDetails.module.css';
import classesCard from '../Card/Card.module.css';

const CardDetails = () => {
  const [, setSearchParams] = useSearchParams();

  const { viewMode, isLoadingDetailingPage } = useAppSelector(
    (state: RootState) => state.CHARACTERS_SLICE
  );
  const dispatch = useAppDispatch();

  const { data, isError } = useGetCharacterItemQuery({
    id: viewMode,
  });

  return (
    <div className={classes.details}>
      {isLoadingDetailingPage && <Preloader />}

      {!isLoadingDetailingPage && !isError && data && (
        <>
          <li className={classesCard.card} data-testid="card-details">
            <p
              className={`${classesCard.card__label} ${
                classesCard[getStatusCharacterColor(data.status)]
              }`}
            >
              {data.status}
            </p>
            <img
              data-testid="card-image"
              className={classesCard.card__image}
              src={data.image}
              alt={data.name}
            />
            <div className={classesCard.card__content}>
              <h2>{data.name}</h2>
              <p>{`Gender: ${data.gender}`}</p>
              <p>{`Species: ${data.species}`}</p>
              <p>{`Location: ${data.location.name}`}</p>
            </div>
          </li>
          <Button
            type="button"
            onClick={() => {
              dispatch(charactersChangeViewMode(''));
              setSearchParams((searchParams) => {
                searchParams.delete(DETAILS_URL_PARAMETER_KEY);
                return searchParams;
              });
            }}
            data-testid="close-details"
          >
            Close
          </Button>
        </>
      )}
    </div>
  );
};

export default CardDetails;
