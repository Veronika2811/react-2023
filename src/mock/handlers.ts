import { HttpResponse, http } from 'msw';

import { BASE_URL } from '@/utils/constants/constants';

export const handlers = [
  http.get(`${BASE_URL}`, () => {
    return HttpResponse.error();
  }),

  http.get(`${BASE_URL}/:id`, ({ params }) => {
    const { details } = params;
    return HttpResponse.json({
      created: '2017-11-04T18:50:21.651Z',
      episode: ['https://rickandmortyapi.com/api/episode/1'],
      gender: 'Male',
      id: details,
      image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
      location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
      },
      name: 'Morty Smith',
      origin: {
        name: 'unknown',
        url: '',
      },
      species: 'Human',
      status: 'Alive',
      type: '',
      url: 'https://rickandmortyapi.com/api/character/2',
    });
  }),
];