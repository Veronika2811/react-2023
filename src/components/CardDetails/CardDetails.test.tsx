import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';
import { expect, it } from 'vitest';
import { screen, fireEvent, render } from '@testing-library/react';

import CardDetails from './CardDetails';
import { DEFAULT_QUERY_PARAMS } from '@/utils/constants/constants';
import RenderWithNextRouter from '@/mock/RenderWithNextRouter';
import { MortyMock } from '@/mock/cardsMock';
import { createMockRouter } from '@/mock/createMockRouter';

describe('CardDetails component', () => {
  it('renders correctly CardDetails component', () => {
    const container = render(
      <RenderWithNextRouter>
        <CardDetails card={MortyMock} />
      </RenderWithNextRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it('should display relevant character data CardDetails component', () => {
    render(
      <RenderWithNextRouter>
        <CardDetails card={MortyMock} />
      </RenderWithNextRouter>
    );

    const cardDetails = screen.getByTestId('card-details');
    expect(cardDetails).toBeInTheDocument();

    const { status, name, gender, species, location } = MortyMock;

    const STATUS = screen.getByText(status);
    expect(STATUS).toBeInTheDocument();

    const NAME = screen.getByText(name);
    expect(NAME).toBeInTheDocument();

    const GENDER = screen.getByText(`Gender: ${gender}`);
    expect(GENDER).toBeInTheDocument();

    const SPECIES = screen.getByText(`Species: ${species}`);
    expect(SPECIES).toBeInTheDocument();

    const LOCATION_NAME = screen.getByText(`Location: ${location.name}`);
    expect(LOCATION_NAME).toBeInTheDocument();
  });

  it('should close the component with character details and leave the existing ones query params', () => {
    const router = createMockRouter({
      query: {
        details: MortyMock.id.toString(),
        name: MortyMock.name,
        page: DEFAULT_QUERY_PARAMS.currentPage,
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });

    render(
      <RouterContext.Provider value={router}>
        <CardDetails card={MortyMock} />
      </RouterContext.Provider>
    );

    const closeButton = screen.getByTestId('close-details');
    fireEvent.click(closeButton);

    expect(router.push).toHaveBeenCalledWith({
      query: {
        name: MortyMock.name,
        page: DEFAULT_QUERY_PARAMS.currentPage,
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  });

  it('should close the character details component and set the query parameters to default', async () => {
    const router = createMockRouter({});

    render(
      <RouterContext.Provider value={router}>
        <CardDetails card={MortyMock} />
      </RouterContext.Provider>
    );

    const closeButton = screen.getByTestId('close-details');
    fireEvent.click(closeButton);

    expect(router.push).toHaveBeenCalledWith({
      query: {
        page: DEFAULT_QUERY_PARAMS.currentPage,
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });
  });
});
