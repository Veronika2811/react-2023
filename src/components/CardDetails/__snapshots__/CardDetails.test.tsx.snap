// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardDetails component > renders correctly CardDetails component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_modal_889028"
      >
        <div
          class="_details_889028"
        >
          <li
            class="_card_f767b2 _card__details_889028"
            data-testid="card-details"
          >
            <p
              class="_card__label_f767b2 _card__label_green_f767b2"
            >
              Alive
            </p>
            <img
              alt="Morty Smith"
              data-nimg="1"
              decoding="async"
              height="250"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
              srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
              style="color: transparent;"
              width="250"
            />
            <div
              class="_card__content_f767b2"
            >
              <h2>
                Morty Smith
              </h2>
              <p>
                Gender: Male
              </p>
              <p>
                Species: Human
              </p>
              <p>
                Location: Citadel of Ricks
              </p>
            </div>
          </li>
          <button
            class="_button_634e7c"
            data-testid="close-details"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_modal_889028"
    >
      <div
        class="_details_889028"
      >
        <li
          class="_card_f767b2 _card__details_889028"
          data-testid="card-details"
        >
          <p
            class="_card__label_f767b2 _card__label_green_f767b2"
          >
            Alive
          </p>
          <img
            alt="Morty Smith"
            data-nimg="1"
            decoding="async"
            height="250"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
            srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
            style="color: transparent;"
            width="250"
          />
          <div
            class="_card__content_f767b2"
          >
            <h2>
              Morty Smith
            </h2>
            <p>
              Gender: Male
            </p>
            <p>
              Species: Human
            </p>
            <p>
              Location: Citadel of Ricks
            </p>
          </div>
        </li>
        <button
          class="_button_634e7c"
          data-testid="close-details"
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
